File: ruby/bob/bob.rb
# Bob responds to remarks
class Bob
  def self.hey(remark)
    remark.gsub!(/\s+/, '')
    response = 'Whatever.'
    response = 'Sure.' if question?(remark)
    response = 'Whoa, chill out!' if shouting?(remark)
    response = 'Fine. Be that way!' if silence?(remark)
    response
  end

  def self.shouting?(remark)
    return false if remark.match(/[a-zA-Z]+/).nil?
    remark.upcase == remark
  end

  def self.question?(remark)
    remark.match(/\?$/)
  end

  def self.silence?(remark)
    remark.empty?
  end
end

module BookKeeping
  VERSION = 1
end

File: javascript/bob/bob.js

const BOB_RESPONSE_TO_QUESTION  = "Sure.";
const BOB_RESPONSE_TO_YELLING   = "Whoa, chill out!";
const BOB_RESPONSE_TO_NOTHING   = "Fine. Be that way!";
const BOB_DEFAULT_RESPONSE      = "Whatever.";

Bob = function(){}

// take question to be anything containing a '?' character
// take yelling to be all caps
// take nothing to be containing no non blank characters

Bob.prototype.hey = function(input){
  if (input ===  input.toUpperCase())     return BOB_RESPONSE_TO_YELLING; 
  if (input.indexOf('?') !== -1)          return BOB_RESPONSE_TO_QUESTION;
  // regex will match string that contains at least one non-space 
  if (input.match(/^\s*$/) !== null) return BOB_RESPONSE_TO_NOTHING;
  return BOB_DEFAULT_RESPONSE;
}

module.exports = Bob;
